### PR TITLE
fix(repeat): add contextual props back

### PR DESCRIPTION
**Pull request by bigopon**
# Pull Request

## 📖 Description

Add back contextual properties to `repeat`, move collection observation logic out of the `Repeat` custom attribute, and employ a new binding type: `IteratorBinding` to handle communication between view model <-> repeat custom attribute

### 🎫 Issues

One of the issue with vCurrent repeat is the observation logic sits inside repeat, and collection handling is done via repeat strategy. This makes it beautiful, but also encourages any implementation of repeat (reorderable repeat, virtual repeat, etc) to re-implement all basic logic again. Try employ a new binding as a communicator to solve this issue.

## 👩‍💻 Reviewer Notes

Introduce new interface `IIteratorBindingTarget` to help communicate changes from view model to the repeat (or any iterator binding target)

## 📑 Test Plan

Added tests for basic scenarios, and mutations scenarios, also with 2 type of value converters: return the same collection instance/value, and returning a new one. (Identity & clone value converters). This covers a lot of common use case, but there could be more (mutate from Array -> Map -> Set -> number);

### CODE REVIEW

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://27957-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://28206-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://28666-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://28674-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://29468-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://29499-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://29637-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://29660-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

File: packages/runtime/src/resources/custom-attributes/repeat.ts [hunk 1]
```diff
@@ -198,9 +190,9 @@ export class Repeat<C extends ObservedCollection = IObservedArray, T extends INo
   }
 
   private processViewsKeyed(indexMap: IndexMap | undefined, flags: LF): void {
+    const oldLength = this.views.length;
     if (indexMap === void 0) {
       if ((this.$controller.state & State.isBoundOrBinding) > 0) {
-        const oldLength = this.views.length;
```
[the remark this answers is not in the document]
Comment (bigopon) [reply packages/runtime/src/resources/custom-attributes/repeat.ts]: In the first iteration, i need to pass old length around, then forgot to move it down.

File: packages/runtime/src/resources/custom-attributes/repeat.ts [hunk 2]
```diff
@@ -528,22 +540,27 @@ export class Repeat<C extends ObservedCollection = IObservedArray, T extends INo
 
     let next: IController;
     let j = seqLen - 1;
-    let i = indexMap.length - 1;
+    let i = newLen - 1;
+    let view: IController;
     for (; i >= 0; --i) {
+      view = views[i];
+      // todo: should not have to always update?
```
[the remark this answers is not in the document]
Comment (bigopon) [reply packages/runtime/src/resources/custom-attributes/repeat.ts]: then i will need to pass old length around, which is why i moved the above up lol.  could do it. Could you help with finishing this piece? the tests are there

**Comment by AureliaEffect** [comment]
JS Framework Benchmark: https://30586-121594173-gh.circle-artifacts.com/0/home/circleci/repo/test/js-framework-benchmark/webdriver-ts-results/table.html

**Pull request merged**